src/Instrumentation/Symfony/src/MessengerInstrumentation.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Instrumentation\Symfony;
+
+use OpenTelemetry\API\Instrumentation\CachedInstrumentation;
+use OpenTelemetry\API\Trace\Span;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\API\Trace\StatusCode;
+use OpenTelemetry\Context\Context;
+use function OpenTelemetry\Instrumentation\hook;
+use OpenTelemetry\SemConv\TraceAttributes;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\Transport\Sender\SenderInterface;
+
+/**
+ * The messenger instrumentation will create an internal span for each message dispatched.
+ *
+ * It is currently not designed to facilitate trace context propagation.
+ * This should be done at the transport level.
+ *
+ * An exception to this will be simple transports like the Doctrine and InMemory transports.
+ *
+ * Caution: MessageBuses can be nested, so we might wand to add
+ * a message stamp to keep track of the parent span.
+ */
+final class MessengerInstrumentation
+{
+    const ATTRIBUTE_MESSENGER_BUS = 'symfony.messenger.bus';
+    const ATTRIBUTE_MESSENGER_MESSAGE = 'symfony.messenger.message';
+    const ATTRIBUTE_MESSENGER_TRANSPORT = 'symfony.messenger.transport';
+
+    public static function register(): void
+    {
+        $instrumentation = new CachedInstrumentation('io.opentelemetry.contrib.php.symfony_messenger');
+
+        /**
+         * MessageBusInterface dispatches messages to the handlers.
+         */
+        hook(
+            MessageBusInterface::class,
+            'dispatch',
+            pre: static function (
+                MessageBusInterface $bus,
+                array $params,
+                string $class,
+                string $function,
+                ?string $filename,
+                ?int $lineno,
+            ) use ($instrumentation): array {
+                /** @var object|Envelope $message */
+                $message = $params[0];
+                $messageClass = \get_class($message);
+
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = $instrumentation
+                    ->tracer()
+                    ->spanBuilder(\sprintf('DISPATCH %s', $messageClass))
+                    ->setSpanKind(SpanKind::KIND_INTERNAL)
+                    ->setAttribute(TraceAttributes::CODE_FUNCTION, $function)
+                    ->setAttribute(TraceAttributes::CODE_NAMESPACE, $class)
+                    ->setAttribute(TraceAttributes::CODE_FILEPATH, $filename)
+                    ->setAttribute(TraceAttributes::CODE_LINENO, $lineno)
+
+                    ->setAttribute(self::ATTRIBUTE_MESSENGER_BUS, $class)
```

Comment:
There are a bunch of messaging-related attributes in SemConv: https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#messaging-attributes - some might be appropriate here, eg `messaging.system`, `messaging.message.id` ?

Reply:
I'm hesitant about the semconv related comments.

In Symfony Messenger, the chain of events is:

Dispatch to Bus -> Send To Transport -> Send to messaging system. 

The actual messaging system will be rabbit, Kafka or the like and there we would add producer kind and so on. I don't want to abuse the semconv as the pieces instrumented here are really internal bits; hence internal spans. The actual instrumentation for interaction with a queue or messaging and the context propagation should be done in a different instrumentation IMHO